wrappers/golang/curves/bls12377/tests/poseidon_test.go

```diff
@@ -0,0 +1,42 @@
+package tests
+
+import (
+	"testing"
+
+	core "github.com/ingonyama-zk/icicle/v2/wrappers/golang/core"
+	cr "github.com/ingonyama-zk/icicle/v2/wrappers/golang/cuda_runtime"
+	bls12_377 "github.com/ingonyama-zk/icicle/v2/wrappers/golang/curves/bls12377"
+	poseidon "github.com/ingonyama-zk/icicle/v2/wrappers/golang/curves/bls12377/poseidon"
+)
+
+func TestPoseidon(t *testing.T) {
+
+	arity := 2
+	numberOfStates := 1
+
+	cfg := poseidon.GetDefaultPoseidonConfig()
+
+	var constants core.PoseidonConstants[bls12_377.ScalarField]
+	ctx, _ := cr.GetDefaultDeviceContext()
+
+	poseidon.InitOptimizedPoseidonConstantsCuda(arity, ctx, &constants) //generate constants
+
+	scalars := bls12_377.GenerateScalars(numberOfStates * arity)
+	scalars[0] = scalars[0].Zero()
+	scalars[1] = scalars[0].Zero()
+
+	scalarsCopy := core.HostSliceFromElements(scalars[:numberOfStates*arity])
+
+	stream, _ := cr.CreateStream()
+
+	var deviceInput core.DeviceSlice
+	scalarsCopy.CopyToDeviceAsync(&deviceInput, stream, true)
+	var deviceOutput core.DeviceSlice
+	deviceOutput.MallocAsync(numberOfStates*scalarsCopy.SizeOfElement(), scalarsCopy.SizeOfElement(), stream)
+
+	poseidon.PoseidonHash(deviceInput, deviceOutput, numberOfStates, arity, &cfg, &constants) //run Hash function
```

Comment:
Should this be running async on the same stream as deviceInput and deviceOutput?

Reply:
Do you mean that this not safe because the Hash function might be run on a slice that is not yet allocated ?

Reply:
I guess that as well but also CopyToDeviceAsync might not have finished copying to the device pointer.
Its also odd to have some portion be async and some not in this case

Reply:
I think using async everywhere is fine. Just make sure it all runs in a single stream